Function Attributes/README.md

```diff
@@ -0,0 +1,524 @@
+```d
+import std.file : FileException, readText;
+import std.stdio : writeln;
+
+void main()
+{
+    try
+    {
+        readText("dummyFile");
+    }
+    catch (FileException e)
+    {
+        writeln("Message:\n", e.msg);
+        writeln("File: ", e.file);
+        writeln("Line: ", e.line);
+        writeln("Stack trace:\n", e.info);
+
+        // Default formatting could be used too
+        // writeln(e);
+    }
+    finally 
+    {
+        writeln("This message will be printed regardless of whether an exception is thrown");
+    }
+}
+```
+
+In D, the exception hierarchy starts with `Throwable`.
+         Throwable (not recommended to catch)
+            ↗                           ↖
+       Exception                         Error (not recommended to catch)
+     ↗         ↖                                        ↗         ↖
+FileException StringException (and others)          RangeError    ...
+
+`Exception` and `Error` are inherited from `Throwable` and are more specific.
+It is not recommended to catch objects of type `Error` and objects of types that are inherited from `Error`.
+That is because they represent system errors that are usually fatal.
+
+The catch blocks are matched in the order in which they are declared, so they must be ordered from the most specific exception types to the most general exception types.
+Given, for example, that the Exception type is the most general, it should be specified at the last catch block, if this is really needed.
+
+```d
+try {
+    // operations about student records that may throw ...
+
+} catch (StudentIdDigitException exc) {
+
+    // an exception that is specifically about errors with
+    // the digits of student ids
+
+} catch (StudentIdException exc) {
+
+    // a more general exception about student ids but not
+    // necessarily about their digits
+
+} catch (StudentRecordException exc) {
+
+    // even more general exception about student records
+
+} catch (Exception exc) {
+
+    // the most general exception that may not be related
+    // to student records
+
+}
+```
+
+The `throw` keyword can be used to deliberately throw exceptions.
+
+Returning to `nothrow`, it guarantees that a function does not emit any exception.
+
+> What is meant here by "any exception" is "any exception that is defined under the Exception hierarchy." A `nothrow` function can still emit exceptions that are under the `Error` hierarchy, which represents irrecoverable error conditions that should preclude the program from continuing its execution.
+
+```d
+import std.stdio;
+
+int add(int lhs, int rhs) nothrow {
+    writeln("adding");    //  compilation ERROR because writeln is not a `nothrow` function
+    return lhs + rhs;
+}
+
+void main() {
+    add(1, 2);
+}
+```
+
+As with purity, the compiler automatically infers whether a template, delegate, or anonymous function is `nothrow`.
+
+*Practice*: make the small program above compile _without_ removing `nothrow` or the `writeln` statement.
```

Comment:
I would also add an exercise here that is a bit more complicated. Also, do you want to maybe provide a file that contains this code so that they don't need to copy paste the code.

Reply:
What others exercises would you propose? 

Ok, will add the code to a file.